### PR TITLE
Task/563 Implement Contributions Widget

### LINKED ISSUE #563: Implement Contributions Widget
_**Note:** Related to #501 

### Problem

At the moment, it's not easy to understand which contributions have been received / made so far unless the user goes to the "Contributions" page. 

### Solution

Captura de ecrã 2019-04-29, às 14 47 30

Add a "widget" that lets the user:
- Keep track of their monthly contributions (whether they are sending or receiving money);
- Keep their stated pledge/need visible (related to issue #501);
- Keep the action to change their pledge / need at hand, as this information might change often;
- Keep a link to non-monetary contributions (help improve feature discoverability).

For detailed mockups of this widget [go here](https://www.figma.com/file/ZR0DRPhhUuHdJtf7PbGAM6tW/Group-Income?node-id=1579%3A12) (desktop, tablet and mobile). You can also see it in the [context of the Dashboard](https://www.figma.com/file/ZR0DRPhhUuHdJtf7PbGAM6tW/Group-Income?node-id=1253%3A14719).

### PR DESCRIPTION
**Pull request by sandrina-p**
Closes #563 

image

image

- When a group is created, the only widget visible is the one to add the income details. Once that is filled, it's the usual flow - Start Inviting and the other widgets.
- Because #757 is still being discussed, I didn't connect the Payments progress bar to the store. For now, it's fake data - it always says "1 out of 3 ". In another PR, after #757, we can finish it along with PayGroup page's sidebar.

### CODE REVIEW

**Review by mmbotelho — COMMENTED** [review]
In terms of layout, everything seems perfect to me! 

However, the data seems to be behaving in a weird way. In this scenario, one member was pledging and the other needed pledges. However, it says that no one is in need of contributions ATM:
Captura de ecrã 2019-12-27, às 16 08 56
Captura de ecrã 2019-12-27, às 16 09 02

**Review by mmbotelho — CHANGES_REQUESTED** [review]
@sandrina-p I tried everything I could think of to replicate the error I mentioned in [my previous comment](https://github.com/okTurtles/group-income-simple/pull/793#pullrequestreview-336824890) but wasn't able to 😢 If you no else can replicate it as well you can ignore it. Maybe it was some error on my part.

I noticed this small visual bug:
Captura de ecrã 2020-01-07, às 18 14 07

Once that is fixed, I'll aprove the PR!

File: frontend/model/state.js
Comment (pieer) [inline frontend/model/state.js:261]: Good point

File: frontend/model/state.js
Comment (pieer) [inline frontend/model/state.js:249]: Add todo refactor "userDisplayName"  in Profile.vue, Welcome.vue (coming with my commit, but I don't want to mess with this code)

File: frontend/views/pages/GroupDashboard.vue
Comment (pieer) [inline frontend/views/pages/GroupDashboard.vue]: Object.keys(currentGroupState.proposals).length > 0
-> should we add this logic into computed property?

File: frontend/model/state.js
Comment (sandrina-p) [reply frontend/model/state.js:249]: done

File: frontend/views/pages/GroupDashboard.vue
Comment (sandrina-p) [reply frontend/views/pages/GroupDashboard.vue]: done

**Comment by sandrina-p** [comment]
@mmbotelho the border bug was fixed. I hope this time is fixed, for real, everywhere.

**Review by mmbotelho — DISMISSED** [review]
Approved! Great job @sandrina-p 🎉

File: frontend/model/contracts/group.js
Comment (taoeffect) [inline frontend/model/contracts/group.js]: What is the reason for this logic? I have a few issues with it:

- It seems unnecessary
- The comment and code assume that we are changing `incomeDetailsType` in this action, which we might not be doing at all

Everything in this file needs to be as clean and minimal as possible, since it will all be "set in stone", and will define a protocol that alternative implementations will have to implement. So anything that isn't necessary should be removed.

File: frontend/model/state.js
Comment (taoeffect) [inline frontend/model/state.js:325]: There are a few issues here:

1. This code belongs in ContributionsWidget.vue, not here. The getters in this file must only be functions that are used throughout the app, and they must be super efficient and compact — i.e. primitive building blocks only. This function is neither highly used throughout the app (it is specific to ContributionsWidget.vue, the use in PaymentsSummary.vue does not need it), nor is it super efficient or compact
2. This function defines and runs code for at least 4 different functions, regardless of whether it's needed or not. It could probably be done more cleanly when moved into ContributionsWidget.vue

Comment (sandrina-p) [reply frontend/model/state.js]: 1. This getter can also be used at `IncomeDetailsGraph.js` and `Contributions.js`. Yes, we only need it 3 times so far, but it's kinda complex and duplicating that code doesn't seem wise.
2. It runs 4 functions, but each one of them has an early return in case it doesn't make sense to perform further logic. I don't see a problem to be as it is for now. When we reuse it at the places mentioned before, if needed, we can evolve it.

Comment (sandrina-p) [reply frontend/model/state.js:325]: 1. This getter can also be used at IncomeDetailsGraph.js and Contributions.js. Yes, we only need it 3 times so far, but it's kinda complex and duplicating that code doesn't seem wise.
2. It runs 4 functions, but each one of them has an early return in case it doesn't make sense to perform further logic. I don't see a problem to be as it is for now. When we reuse it at the places mentioned before, if needed, we can evolve it.

File: frontend/model/contracts/group.js
Comment (sandrina-p) [reply frontend/model/contracts/group.js]: Removed, it's not necessary indeed. However, I still don't like the idea of leaving "traces" of stale data, there's no reason to keep a value that doesn't make sense anymore.

Comment (taoeffect) [reply frontend/model/contracts/group.js]: You might be able to fix that in the UI portion of the app

Comment (sandrina-p) [reply frontend/model/contracts/group.js]: > You might be able to fix that in the UI portion of the app

hmm but there's nothing to fix ou the UI side. this is a state optimization, not UI bug.

Comment (taoeffect) [reply frontend/model/contracts/group.js]: There is definitely something that can be fixed on the UI side: reset the value. The effect to the user will be the same, while the logic in our contract chains will be kept to a minimum

Comment (sandrina-p) [reply frontend/model/contracts/group.js]: I think I'm not understanding... reset the value where exactly? At the moment is all working fine.

Comment (taoeffect) [reply frontend/model/contracts/group.js]: In the modal for the income details. If you're worried about a stale value appearing there, that's where you can reset it.

File: frontend/views/pages/GroupDashboard.vue
Comment (taoeffect) [inline frontend/views/pages/GroupDashboard.vue]: The organization here needs to be improved. See my [two comments](https://github.com/okTurtles/group-income-simple/pull/789/files#diff-7eb644607ce1945a00d85cf4544ada76) to Pierre in `GroupDashboard.vue` about how the logic in this file should be organized. I think we should wait until Pierre's PR is merged before merging this one so that you can resolve the merge conflicts, but you might also want to clean some of this up beforehand too, or at least think about how to go about it.

From a pure programming perspective, it's best to avoid checking the same condition over and over, in this case you have a lot of checks against `hasIncomeDetails`, and ideally there would only ever be one (plus an optional `v-else`).

File: frontend/model/contracts/group.js
Comment (taoeffect) [inline frontend/model/contracts/group.js]: This change doesn't seem to be needed?

File: frontend/model/state.js
Comment (taoeffect) [reply frontend/model/state.js:325]: > This getter can also be used at IncomeDetailsGraph.js and Contributions.js. Yes, we only need it 3 times so far, but it's kinda complex and duplicating that code doesn't seem wise.

I can't find those files in the project... Do you mean `IncomeDetails.vue` and `Contributions.vue`?

I agree that code duplication isn't a good idea. However, if you're going to create code in order to de-duplicate code, then go ahead and do so by modifying those files n this PR.

We still need to address the issue of this function being very large, and running all of the functions regardless of what data is being called.

There has to be a cleaner and more modular way to do this, and a way in which code is run only when it needs to be run. Do you think you can see if you can clean this up a bit while DRYing the other files too? Maybe also finding another file to put this code in (perhaps creating it if necessary)? The thing is, `state.js` is already bigger than it should be. You've got time since Pierre's PR needs to be merged first, and if you'd like I can lend a hand after Pierre's PR is merged.

**Review by taoeffect — CHANGES_REQUESTED** [review]
(This isn't a full review but a partial one.)

File: frontend/views/pages/GroupDashboard.vue
Comment (sandrina-p) [reply frontend/views/pages/GroupDashboard.vue]: The checks for `hasIncomeDetails` are essential. If you see a simpler way to avoid those checks, please let me know

File: frontend/model/contracts/group.js
Comment (sandrina-p) [reply frontend/model/contracts/group.js]: line removed

File: frontend/model/state.js
Comment (sandrina-p) [reply frontend/model/state.js:325]: Yes, I did mean .vue files, not .js.

> and a way in which code is run only when it needs to be run.
This code only runs once, when it's checked.

 I'll take a closer look on these as you suggested and let you know when it's ready.

Comment (sandrina-p) [reply frontend/model/state.js:325]: done @taoeffect. I refactored `Contributions.vue` to use the getter `ourContributionsSummary` and it needed all the 4 functions as I was expecting. So, in summary, this getter is used twice and 100% needed, no code could be removed.

I strongly believe that the right place for `ourContributionsSummary` is as a getter, as it is. The main advantage of using getters instead of a "standard" util function, is that a getter is only run once in the entire app, even if called multiple times (in our case, it's called twice - on the widget and on the contributions page).

> The thing is, state.js is already bigger than it should be.

Totally agree, but this isn't a problem of `ourContributionsSummary`. I mentioned some time ago that `state.js` as it is today, is not scalable. We should take better advantage of Vuex modules, and split the `state.js` into modules, [(its a common pattern for large vuex apps)](https://medium.com/3yourmind/large-scale-vuex-application-structures-651e44863e2f). Well, this is a very valid point that requires further discussion, but I don't see this directly related to this PR. If you want, I can create an issue for us to discuss state there.

File: frontend/assets/style/components/_links.scss
Comment (taoeffect) [inline frontend/assets/style/components/_links.scss:2]: These comments are super helpful 👍

File: frontend/model/state.js
Comment (taoeffect) [inline frontend/model/state.js:252]: This is a great example of a getter that definitely belongs in this file, great job! 👍

File: frontend/views/components/PageSection.vue
Comment (taoeffect) [inline frontend/views/components/PageSection.vue:6]: Is this a @mmbotelho sanctioned change? :)

File: frontend/views/pages/GroupDashboard.vue
Comment (taoeffect) [inline frontend/views/pages/GroupDashboard.vue]: `groupProfile` is only used once in this file, and it's used to do exactly what `ourGroupProfile` does, so please delete this and replace line 63 with `this.ourGroupProfile`

File: frontend/model/state.js
Comment (taoeffect) [inline frontend/model/state.js]: Use `getters.ourGroupProfile`

File: frontend/model/state.js
Comment (taoeffect) [reply frontend/model/state.js:325]: > Well, this is a very valid point that requires further discussion, but I don't see this directly related to this PR. If you want, I can create an issue for us to discuss state there.

OK, that sounds reasonable, please go ahead and create an issue for this, and link to this thread (leaving it unresolved)

File: frontend/views/components/CalloutCard.vue
Comment (taoeffect) [inline frontend/views/components/CalloutCard.vue:39]: @pieer does this change make sense to you?

File: frontend/model/state.js
Comment (taoeffect) [inline frontend/model/state.js]: These are contributions, not non-contributions, so rename to `nonMonetaryContributionsOf`

**Review by taoeffect — CHANGES_REQUESTED** [review]
My review is almost, but not quite finished, there are still some parts I haven't checked. Leaving comments for what I was able to get to today though!

File: frontend/views/components/CalloutCard.vue
Comment (pieer) [reply frontend/views/components/CalloutCard.vue:39]: I confirm it make sense :) It's a design choice that respect @mmbotelho design

NB: Sandrina created the CalloutCard, that part concern the SVG icon on the left of the callout text, not the graph

File: frontend/views/components/PageSection.vue
Comment (sandrina-p) [reply frontend/views/components/PageSection.vue:6]: No, it's just a matter of having the correct heading structure. This component is used as sections of the pages, so the page title is an h1, which means each card is an h2. if it was h3, we would have a heading gap and that shouldn't happen. Here's a screenshot for you to better understand:

image

File: frontend/model/state.js
Comment (sandrina-p) [reply frontend/model/state.js]: It was a typo, I actually wanted to write `nonMonetaryContributionsOf `! 😶

File: frontend/model/state.js
Comment (sandrina-p) [reply frontend/model/state.js]: Done. https://github.com/okTurtles/group-income-simple/issues/804

Comment (sandrina-p) [reply frontend/model/state.js:325]: Done. #804

File: frontend/views/components/PageSection.vue
Comment (taoeffect) [reply frontend/views/components/PageSection.vue:6]: So I assume that means h2 and h3 have the same style...?

Comment (sandrina-p) [reply frontend/views/components/PageSection.vue:6]: the style is applied by the class `.is-title-X`, it has nothing to do with the heading tag

Comment (taoeffect) [reply frontend/views/components/PageSection.vue:6]: ok, so in other words, does that mean the heading tags should never be used without an `.is-title-X` marker...? Usually the h1, h2, h3 tags determine the style/size by themselves. Why don't they do that in our project, and why do they need a `.is-title-X` for it? (Is this something that can be explained in the Style Guide CSS section? If so can you create an issue for it?)

Comment (sandrina-p) [reply frontend/views/components/PageSection.vue:6]: That was a change made a few PR ago: https://github.com/okTurtles/group-income-simple/pull/777 Basically we did it to support correctly responsive headers.

> Usually the h1, h2, h3 tags determine the style/size by themselves. 

This is a misconception/bad practice in terms of accessibility. Markup (semantics) should be decoupled from styling. We can style something to look like a heading and not be an actual heading or vice-versa.

>  (Is this something that can be explained in the Style Guide CSS section? If so can you create an issue for it?)

Yes, this is the [best/simpler explanation of why](https://stackoverflow.com/a/19166926/4737729)

Comment (taoeffect) [reply frontend/views/components/PageSection.vue:6]: OK thanks, I've gone ahead and created issue #805 for the documentation about this

File: frontend/views/pages/Contributions.vue
Comment (taoeffect) [inline frontend/views/pages/Contributions.vue:100]: Nice catch! 👍

File: frontend/views/pages/Contributions.vue
Comment (taoeffect) [inline frontend/views/pages/Contributions.vue]: DRY: already defined in `state.js`

File: frontend/views/pages/GroupDashboard.vue
Comment (taoeffect) [inline frontend/views/pages/GroupDashboard.vue]: DRY: use `hasProposals`

EDIT: re: `master` conflict, this change should be combined with what's in `master`, i.e. the indentation should be decreased (as in Pierre's recent quickfix for #806), but we should also only show it if there are proposals to show

File: frontend/views/pages/Page.vue
Comment (taoeffect) [inline frontend/views/pages/Page.vue]: Does this need a `.is-title-2` on it? If not, can you add a comment explaining why not? (And maybe add a comment to #805 to use this as an example of when not to use it?)

File: frontend/views/containers/Welcome.vue
Comment (taoeffect) [inline frontend/views/containers/Welcome.vue]: This should be the same as the filename. So either this needs to be `Welcome` or the file should be renamed

File: frontend/views/containers/contributions/ContributionsWidget.vue
Comment (taoeffect) [inline frontend/views/containers/contributions/ContributionsWidget.vue]: Should the space after `{br_}` be there?

**Review by taoeffect — CHANGES_REQUESTED** [review]
Great job @sandrina-p! Minor changes requested. I think this is also ready for @mmbotelho to review to verify everything works per her designs. I still need to do my own tests as well and will do that soon, probably after the next set of changes (or earlier)

File: test/cypress/integration/group-payments.spec.js
Comment (taoeffect) [inline test/cypress/integration/group-payments.spec.js:150]: ~~Re tests: do you think you can add a couple tests to test that this new widget works?~~

Oh, it's not hooked up to the state yet. But does that mean there's nothing at all that can be tested?

File: frontend/views/pages/GroupDashboard.vue
Comment (pieer) [reply frontend/views/pages/GroupDashboard.vue]: I think merging master might resolve the problem here

Comment (taoeffect) [reply frontend/views/pages/GroupDashboard.vue]: Oh yeah, this is already outdated, I'll resolve this thread

File: frontend/views/pages/Page.vue
Comment (pieer) [reply frontend/views/pages/Page.vue]: YEs, to stay consistent we need the class to the title now

Comment (sandrina-p) [reply frontend/views/pages/Page.vue]: No, it doesn't. This is a heading only for screen readers, so they know what's this sidebar about. That's the purpose of `.sr-only`.

File: frontend/views/containers/Welcome.vue
Comment (sandrina-p) [reply frontend/views/containers/Welcome.vue]: File renamed to `AddIncomeDetailsWidget`.

File: frontend/views/containers/contributions/ContributionsWidget.vue
Comment (sandrina-p) [reply frontend/views/containers/contributions/ContributionsWidget.vue]: ~It's irrelevant... it has no influence on the layout. I would keep the space before and after fo readability.~

Edited: It doesn't make a difference on the layout, but on testing it makes. Ideally, it should have only 1 space between one word and another. So in this case, I added the space before {br_} and removed the one after. That way we can assert with one space only:
- "You are pledging $100" (correct - one space)
- "You are pledging &nbsp;$100" (incorrect - double space)

File: test/cypress/integration/group-payments.spec.js
Comment (sandrina-p) [reply test/cypress/integration/group-payments.spec.js:150]: Yes, it is hooked up to the store! The payment progress is the only part not hooked yet. But nevertheless, I just added some tests to Contributions Widget as part of `group-payments.spec.js`!

File: frontend/views/pages/Contributions.vue
Comment (taoeffect) [reply frontend/views/pages/Contributions.vue]: Still here :)

File: frontend/views/pages/Contributions.vue
Comment (taoeffect) [inline frontend/views/pages/Contributions.vue]: I like how the other file called this function `withCurrency`, can we make this also clearer by naming it `withCurrency`?

File: frontend/views/pages/Page.vue
Comment (taoeffect) [inline frontend/views/pages/Page.vue:25]: Can you update our documentation using this as an example of when to use `h2` without also adding an `.is-title-X` and explaining why?

File: frontend/views/pages/Page.vue
Comment (taoeffect) [inline frontend/views/pages/Page.vue:8]: Is there a reason this was deleted?

Comment (sandrina-p) [reply frontend/views/pages/Page.vue:8]: Yes. This image doesn't add any extra value to the content, it's only decorative. This is a title and having `{{ title }}` is enough. There are no benefits for the SR (screen reader) to read this image to the user, so in this case we just leave an empty alt and the SR will ignore the image.

**Review by taoeffect — APPROVED** [review]
Great job @sandrina-p! Excellent work! 😄  🎉

**Pull request merged**
